File: implementations/csharp/client/src/Specus.Client/Cli/LocalUi.cs
using System.Diagnostics;
using System.Net;
using System.Net.Sockets;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using Specus.Client.Configuration;
using Specus.Client.Updates;

namespace Specus.Client.Cli;

/// <summary>
/// Deliberately narrow HTTP/1.1 endpoint: one request per connection, fixed-length JSON only,
/// no upgrades, chunking, proxies or filesystem routing. Uses loopback sockets, not HTTP.sys
/// (which would require a Windows URL reservation) or an extra ASP.NET shared runtime.
/// </summary>
internal sealed class LocalUi(ClientCliOptions options, string configPath)
{
    internal sealed class Failure(int status, string message) : Exception(message) { internal int Status => status; }
    private readonly UiRuntime _runtime = new();
    private readonly SemaphoreSlim _operations = new(1);
    private readonly object _authGate = new();
    private readonly Dictionary<string, DateTimeOffset> _sessions = [];
    private string _code = "", _origin = "";
    private DateTimeOffset _codeExpiry;
    private static string Random() => Convert.ToHexStringLower(RandomNumberGenerator.GetBytes(32));
    private void PrintCode()
    {
        lock (_authGate) { _code = Random(); _codeExpiry = DateTimeOffset.UtcNow.AddMinutes(5);
            Console.WriteLine("一次性连接码（5 分钟有效；按 Enter 生成新码）：\n" + _code); }
    }
    private object Session(JsonObject body)
    {
        lock (_authGate)
        {
            var now = DateTimeOffset.UtcNow;
            if (_code == "" || now >= _codeExpiry || body["code"]?.GetValue<string>() != _code) throw new Failure(401, "连接码无效或已过期，请在终端按 Enter 获取新码");
            foreach (var expired in _sessions.Where(s => s.Value <= now).Select(s => s.Key).ToArray()) _sessions.Remove(expired);
            if (_sessions.Count >= 8) throw new Failure(429, "会话数已达上限，请重启本地页面服务");
            _code = ""; var token = Random(); _sessions[token] = now.AddHours(8); return new { schemaVersion = 1, token };
        }
    }
    private void Authenticate(string bearer)
    {
        lock (_authGate)
            if (!bearer.StartsWith("Bearer ", StringComparison.Ordinal) || !_sessions.TryGetValue(bearer[7..], out var expiry) || expiry <= DateTimeOffset.UtcNow)
                throw new Failure(401, "请使用终端中的一次性连接码打开工作台");
    }
    private List<object> Others()
    {
        CliState.CheckPrivate(CliState.Root);
        var result = new List<object>(); int count = 0;
        foreach (var path in Directory.EnumerateFiles(CliState.Root, "*.json"))
        {
            if (++count > 256) throw new Failure(409, "本地状态文件过多，请检查状态目录");
            try
            {
                CliState.CheckPrivate(path); if (new FileInfo(path).Length > 1024 * 1024) continue;
                var data = JsonNode.Parse(File.ReadAllText(path))!;
                int pid = data["pid"]!.GetValue<int>(); long age = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - data["updatedAtUnixMs"]!.GetValue<long>();
                if (pid == Environment.ProcessId || age is < 0 or > 5000 || data["schemaVersion"]!.GetValue<int>() != 1 || !string.Equals(data["configPath"]!.GetValue<string>(), configPath,
                    OperatingSystem.IsWindows() ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal)) continue;
                using var process = Process.GetProcessById(pid); if (process.HasExited) continue;
                result.Add(new { pid, phase = data["phase"]!.GetValue<string>(), controlAuthenticated = data["controlAuthenticated"]!.GetValue<bool>(), businessReady = data["businessReady"]!.GetValue<bool>(), readOnly = true });
            }
            catch (Exception e) when (e is FileNotFoundException or JsonException or ArgumentException or InvalidOperationException or NullReferenceException) { }
        }
        return result;
    }
    private async Task<object> ApiAsync(string path, string method, JsonObject body, CancellationToken cancellation)
    {
        await _operations.WaitAsync(cancellation);
        try
        {
            if (path == "/api/config" && method == "GET") return UiConfig.View(configPath);
            if (path == "/api/status" && method == "GET")
            {
                List<object> others = []; string warning = "";
                try { others = Others(); } catch (Exception) { warning = "无法安全读取其他实例状态，连接操作已禁用"; }
                return new { schemaVersion = 1, implementation = "dotnet", version = ClientVersion.Current, configPath,
                    runtime = _runtime.Snapshot(), otherInstances = others, instanceWarning = warning };
            }
            if ((path is "/api/config/save" or "/api/config/validate") && method == "POST")
            {
                var warnings = new List<string>(); var bytes = UiConfig.Prepare(configPath, body, warnings);
                bool save = path.EndsWith("/save", StringComparison.Ordinal); if (save) UiConfig.Save(configPath, body["revision"]!.GetValue<string>(), bytes);
                return new { schemaVersion = 1, saved = save, offline = true, warnings };
            }
            if (path == "/api/connection" && method == "POST")
            {
                string action = body["action"]?.GetValue<string>() ?? "";
                if (action is not ("start" or "stop" or "restart")) throw new Failure(400, "未知连接操作");
                if (action == "stop") await _runtime.StopAsync();
                else
                {
                    if (Others().Count > 0) throw new Failure(409, "已有其他 CLI 实例运行，本页只读展示且不会接管");
                    var snapshot = UiConfig.Read(configPath); UiConfig.CheckRevision(snapshot, body["revision"]?.GetValue<string>() ?? ""); UiConfig.Parse(snapshot.Bytes);
                    var config = SpecusClientConfigLoader.Parse(new UTF8Encoding(false, true).GetString(snapshot.Bytes), configPath);
                    config.AutoUpdate = false; config.UpdateEnabled = false;
                    if (action == "restart") await _runtime.StopAsync();
                    _runtime.Start(config, configPath, snapshot.Revision, options.LoginTimeout);
                }
                return new { schemaVersion = 1, accepted = true };
            }
            if (path is "/api/config" or "/api/status" or "/api/config/save" or "/api/config/validate" or "/api/connection") throw new Failure(405, "此接口不支持该方法");
            throw new Failure(404, "接口不存在");
        }
        finally { _operations.Release(); }
    }
    private async Task HandleAsync(TcpClient socket, CancellationToken shutdown)
    {
        using (socket)
        using (var deadline = CancellationTokenSource.CreateLinkedTokenSource(shutdown))
        {
            deadline.CancelAfter(TimeSpan.FromSeconds(25)); var ct = deadline.Token; var stream = socket.GetStream();
            int status = 200; string type = "application/json; charset=utf-8"; byte[] result;
            try
            {
                // A separate total read budget prevents slow headers/bodies from retaining a slot.
                using var readDeadline = CancellationTokenSource.CreateLinkedTokenSource(ct); readDeadline.CancelAfter(TimeSpan.FromSeconds(10));
                byte[] header = new byte[16384], single = new byte[1]; int size = 0;
                while (true)
                {
                    if (size == header.Length) throw new Failure(400, "请求头过大");
                    if (await stream.ReadAsync(single, readDeadline.Token) == 0) return;
                    byte b = single[0]; if (b > 126 || b < 32 && b is not (9 or 10 or 13)) throw new Failure(400, "无效请求头");
                    header[size++] = b;
                    if (size >= 4 && header.AsSpan(size - 4, 4).SequenceEqual("\r\n\r\n"u8)) break;
                }
                string[] lines = Encoding.ASCII.GetString(header, 0, size - 4).Split("\r\n");
                string[] first = lines[0].Split(' ');
                if (first.Length != 3 || first[2] != "HTTP/1.1") throw new Failure(400, "仅支持 HTTP/1.1 请求");
                var headers = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
                foreach (string line in lines.Skip(1))
                {
                    int colon = line.IndexOf(':'); if (colon <= 0) throw new Failure(400, "无效请求头");
                    string name = line[..colon];
                    if (name.Any(c => !char.IsAsciiLetterOrDigit(c) && c != '-')) throw new Failure(400, "无效请求头");
                    if (!headers.TryAdd(name, line[(colon + 1)..].Trim(' ', '\t'))) throw new Failure(400, "重复请求头");
                }
                string Get(string key) => headers.GetValueOrDefault(key, "");
                string method = first[0], path = first[1];
                if (Get("Host") != _origin[7..]) throw new Failure(403, "Host 不匹配");
                if (headers.ContainsKey("Origin") && Get("Origin") != _origin || Get("Sec-Fetch-Site") == "cross-site") throw new Failure(403, "仅允许本地页面访问");
                if (method is not ("GET" or "POST")) throw new Failure(405, "不支持该方法");
                if (path.Contains('?') || path.Contains('%') || !path.StartsWith('/')) throw new Failure(400, "不支持查询参数或编码路径");
                if (headers.ContainsKey("Transfer-Encoding") || headers.ContainsKey("Expect") || headers.ContainsKey("Upgrade")) throw new Failure(400, "不支持流式请求或协议升级");
                int length = 0;
                if (headers.TryGetValue("Content-Length", out var rawLength) && (rawLength.Length == 0 || rawLength.Any(c => c is < '0' or > '9') || !int.TryParse(rawLength, out length) || length > 65536)) throw new Failure(400, "请求体过大或长度无效");
                if (method == "POST")
                {
                    if (Get("Origin") != _origin || Get("X-Specus-UI") != "1") throw new Failure(403, "写入操作必须来自本地页面");
                    if (!Get("Content-Type").Split(';')[0].Trim().Equals("application/json", StringComparison.OrdinalIgnoreCase)) throw new Failure(415, "请求必须是 JSON");
                }
                byte[] bytes = new byte[length]; await stream.ReadExactlyAsync(bytes, readDeadline.Token);
                if (method == "GET" && path is "/" or "/app.js" or "/app.css")
                {
                    string name = path == "/" ? "index.html" : path[1..];
                    using var asset = typeof(LocalUi).Assembly.GetManifestResourceStream("local-ui." + name) ?? throw new Failure(404, "页面资源缺失");
                    using var output = new MemoryStream(); await asset.CopyToAsync(output, ct); result = output.ToArray();
                    type = path.EndsWith(".js") ? "text/javascript; charset=utf-8" : path.EndsWith(".css") ? "text/css; charset=utf-8" : "text/html; charset=utf-8";
                }
                else
                {
                    JsonObject body = [];
                    if (method == "POST")
                    { try { body = JsonNode.Parse(bytes) as JsonObject ?? throw new JsonException(); } catch (JsonException) { throw new Failure(400, "无效 JSON 请求"); } }
                    object data;
                    if (path == "/api/session" && method == "POST") data = Session(body);
                    else { Authenticate(Get("Authorization")); data = await ApiAsync(path, method, body, ct); }
                    result = JsonSerializer.SerializeToUtf8Bytes(data);
                }
            }
            catch (OperationCanceledException) { return; }
            catch (Exception error)
            {
                status = error is Failure failure ? failure.Status : 422;
                result = JsonSerializer.SerializeToUtf8Bytes(new { schemaVersion = 1, error = error is Failure ? error.Message : "操作失败，请检查配置、权限和本地状态；原始诊断不会返回浏览器" });
            }
            string response = $"HTTP/1.1 {status} Response\r\nContent-Type: {type}\r\nContent-Length: {result.Length}\r\nConnection: close\r\nCache-Control: no-store\r\nX-Content-Type-Options: nosniff\r\nReferrer-Policy: no-referrer\r\nX-Frame-Options: DENY\r\nPermissions-Policy: camera=(), microphone=(), geolocation=()\r\nContent-Security-Policy: default-src 'none'; script-src 'self'; style-src 'self'; connect-src 'self'; img-src 'self'; base-uri 'none'; frame-ancestors 'none'; form-action 'none'\r\n\r\n";
            try
            {
                await stream.WriteAsync(Encoding.ASCII.GetBytes(response), ct); await stream.WriteAsync(result, ct);
                socket.Client.Shutdown(SocketShutdown.Send);
                // On Windows, closing with unread rejected request bytes resets the socket and
                // can discard the error response. Drain only a small bounded tail after FIN.
                using var drain = CancellationTokenSource.CreateLinkedTokenSource(ct); drain.CancelAfter(TimeSpan.FromMilliseconds(200));
                byte[] tail = new byte[8192]; int remaining = 65537;
                while (remaining > 0)
                { int read = await stream.ReadAsync(tail.AsMemory(0, Math.Min(tail.Length, remaining)), drain.Token); if (read == 0) break; remaining -= read; }
            }
            catch (Exception error) when (error is IOException or OperationCanceledException) { }
        }
    }
    internal static async Task<int> RunAsync(ClientCliOptions options, string configPath)
    {
        Console.OutputEncoding = new UTF8Encoding(false);
        string? ownedLock = null; string stage = "私有状态目录"; var ui = new LocalUi(options, configPath);
        using var shutdown = new CancellationTokenSource(); var listener = new TcpListener(IPAddress.Loopback, options.UiPort);
        using var slots = new SemaphoreSlim(16); var pending = new List<Task>();
        ConsoleCancelEventHandler cancel = (_, e) => { e.Cancel = true; shutdown.Cancel(); };
        try
        {
            string path = Path.Combine(CliState.EnsureRoot(), CliState.Prefix(configPath)[7..] + "ui.lock");
            stage = "同配置管理锁（先确认原管理进程已退出，再清理遗留锁）";
            using (var file = new FileStream(path, FileMode.CreateNew, FileAccess.Write, FileShare.Read))
            { ownedLock = path; file.Write(Encoding.ASCII.GetBytes(Environment.ProcessId.ToString())); }
            stage = "本机端口（检查 --port 是否被占用，或使用 --port 0）";
            listener.Start(32); ui._origin = "http://127.0.0.1:" + ((IPEndPoint)listener.LocalEndpoint).Port;
            stage = "管理服务";
            Console.CancelKeyPress += cancel;
            Console.WriteLine("本地管理页面：" + ui._origin + "\n配置：" + configPath + "\nCtrl+C 退出管理服务及本页连接。"); ui.PrintCode();
            if (!options.NoOpen)
            {
                try
                {
                    var start = new ProcessStartInfo(OperatingSystem.IsWindows() ? "rundll32" : OperatingSystem.IsMacOS() ? "open" : "xdg-open") { UseShellExecute = false, CreateNoWindow = true, WindowStyle = ProcessWindowStyle.Hidden, RedirectStandardOutput = true, RedirectStandardError = true };
                    if (OperatingSystem.IsWindows()) start.ArgumentList.Add("url.dll,FileProtocolHandler");
                    start.ArgumentList.Add(ui._origin + "/#" + ui._code); using var browser = Process.Start(start);
                }
                catch (Exception) { Console.Error.WriteLine("未能打开浏览器，请手动打开本地地址并输入连接码。"); }
            }
            _ = Task.Run(() => { while (Console.ReadLine() is not null && !shutdown.IsCancellationRequested) ui.PrintCode(); });
            while (!shutdown.IsCancellationRequested)
            {
                var socket = await listener.AcceptTcpClientAsync(shutdown.Token);
                if (!slots.Wait(0)) { socket.Dispose(); continue; }
                pending.RemoveAll(task => task.IsCompleted);
                pending.Add(Task.Run(async () => { try { await ui.HandleAsync(socket, shutdown.Token); } finally { slots.Release(); } }));
            }
            return 0;
        }
        catch (OperationCanceledException) when (shutdown.IsCancellationRequested) { return 0; }
        catch (Exception) { Console.Error.WriteLine("无法启动本地页面：" + stage + "。请检查对应资源及当前用户权限。"); return 2; }
        finally
        {
            shutdown.Cancel(); listener.Stop(); Console.CancelKeyPress -= cancel;
            try { await Task.WhenAll(pending); await ui._runtime.StopAsync(); }
            catch (Exception) { Console.Error.WriteLine("连接清理未能按时完成，请确认客户端进程已退出。"); }
            if (ownedLock is not null) File.Delete(ownedLock);
        }
    }
}

File: implementations/java/client/src/main/java/com/theshuai/specusclient/cli/LocalUi.java
package com.theshuai.specusclient.cli;

import com.fasterxml.jackson.databind.JsonNode;
import io.netty.bootstrap.ServerBootstrap;
import io.netty.channel.*;
import io.netty.channel.nio.NioIoHandler;
import io.netty.channel.socket.SocketChannel;
import io.netty.channel.socket.nio.NioServerSocketChannel;
import io.netty.handler.codec.http.*;
import io.netty.handler.timeout.ReadTimeoutHandler;
import io.netty.buffer.Unpooled;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.security.SecureRandom;
import java.util.*;
import java.util.concurrent.*;

/** Loopback-only, bounded HTTP adapter. The public shell contains no state or credentials. */
public final class LocalUi {
    static final class Failure extends RuntimeException {
        final int status;
        Failure(int status,String message) { super(message);this.status=status; }
    }
    private final ClientCli.Options options;
    private final String version;
    private final UiRuntime runtime=new UiRuntime();
    private final Map<String,Long> sessions=new HashMap<>();
    private String code="", origin="";
    private long codeExpiry;
    private final Object operations=new Object();
    private LocalUi(ClientCli.Options options,String version) {this.options=options;this.version=version;}
    private static String random() {byte[] bytes=new byte[32];new SecureRandom().nextBytes(bytes);return HexFormat.of().formatHex(bytes);}
    private synchronized String rotate() {code=random();codeExpiry=System.currentTimeMillis()+300_000;return code;}
    private void printCode() {System.out.println("一次性连接码（5 分钟有效；按 Enter 生成新码）：\n"+rotate());System.out.flush();}
    private synchronized Object session(JsonNode body) {
        long now=System.currentTimeMillis();
        if(now>=codeExpiry||code.isEmpty()||!code.equals(body.path("code").asText())) throw new Failure(401,"连接码无效或已过期，请在终端按 Enter 获取新码");
        sessions.entrySet().removeIf(e -> e.getValue()<=now);
        if(sessions.size()>=8) throw new Failure(429,"会话数已达上限，请重启本地页面服务");
        code="";String token=random();sessions.put(token,now+8*60*60*1000L);
        return Map.of("schemaVersion",1,"token",token);
    }
    private synchronized void authenticate(String bearer) {
        if(bearer==null||!bearer.startsWith("Bearer ")||sessions.getOrDefault(bearer.substring(7),0L)<=System.currentTimeMillis())
            throw new Failure(401,"请使用终端中的一次性连接码打开工作台");
    }
    private List<Object> others() throws Exception {
        var result=new ArrayList<Object>();CliState.checkPrivate(CliState.root());
        try(var files=Files.newDirectoryStream(CliState.root(),"*.json")) {
            int count=0;
            for(var file:files) {
                if(++count>256) throw new Failure(409,"本地状态文件过多，请检查状态目录");
                try {
                    CliState.checkPrivate(file);if(Files.size(file)>1024*1024)continue;
                    var data=UiConfig.JSON.readTree(Files.readString(file));
                    long pid=data.path("pid").asLong(),age=System.currentTimeMillis()-data.path("updatedAtUnixMs").asLong();
                    String config=data.path("configPath").asText();
                    boolean matches=System.getProperty("os.name").startsWith("Windows")?config.equalsIgnoreCase(options.config().toString()):config.equals(options.config().toString());
                    if(pid==ProcessHandle.current().pid()||age<0||age>5000||!matches||data.path("schemaVersion").asInt()!=1||ProcessHandle.of(pid).filter(ProcessHandle::isAlive).isEmpty())continue;
                    result.add(Map.of("pid",pid,"phase",data.path("phase").asText(),"controlAuthenticated",data.path("controlAuthenticated").asBoolean(),
                            "businessReady",data.path("businessReady").asBoolean(),"readOnly",true));
                } catch(NoSuchFileException|com.fasterxml.jackson.core.JsonProcessingException ignored) { }
            }
        }
        return result;
    }
    private Object api(String path,String method,JsonNode body) throws Exception {
        synchronized(operations) {
            if(path.equals("/api/config")&&method.equals("GET")) return UiConfig.view(options.config());
            if(path.equals("/api/status")&&method.equals("GET")) {
                List<Object> others=List.of();String warning="";
                try {others=others();} catch(Exception error) {warning="无法安全读取其他实例状态，连接操作已禁用";}
                return Map.of("schemaVersion",1,"implementation","java","version",version,"configPath",options.config().toString(),
                        "runtime",runtime.snapshot(),"otherInstances",others,"instanceWarning",warning);
            }
            if(Set.of("/api/config/save","/api/config/validate").contains(path)&&method.equals("POST")) {
                var warnings=new ArrayList<String>();String text=UiConfig.prepare(options.config(),body,warnings);
                boolean save=path.endsWith("/save");if(save)UiConfig.save(options.config(),body.path("revision").asText(),text);
                return Map.of("schemaVersion",1,"saved",save,"offline",true,"warnings",warnings);
            }
            if(path.equals("/api/connection")&&method.equals("POST")) {
                String action=body.path("action").asText();
                if(!Set.of("start","stop","restart").contains(action))throw new Failure(400,"未知连接操作");
                if(action.equals("stop")) runtime.stop();
                else {
                    if(!others().isEmpty()) throw new Failure(409,"已有其他 CLI 实例运行，本页只读展示且不会接管");
                    var snapshot=UiConfig.read(options.config());UiConfig.checkRevision(snapshot,body.path("revision").asText());
                    UiConfig.document(snapshot.text());var config=ClientCli.parse(snapshot.text(), ignored -> { });config.setUpdateCheckEnabled(false);
                    if(action.equals("restart")) runtime.stop();
                    runtime.start(config,options.config(),snapshot.revision(),options.loginTimeoutSeconds());
                }
                return Map.of("schemaVersion",1,"accepted",true);
            }
            if(Set.of("/api/config","/api/status","/api/config/save","/api/config/validate","/api/connection").contains(path)) throw new Failure(405,"此接口不支持该方法");
            throw new Failure(404,"接口不存在");
        }
    }
    private void handle(ChannelHandlerContext ctx,String path,String method,HttpHeaders headers,byte[] bytes) {
        int status=200;byte[] result;String type="application/json; charset=utf-8";
        try {
            if(!origin.substring(7).equals(headers.get("Host"))||headers.getAll("Host").size()!=1)throw new Failure(403,"Host 不匹配");
            String requestOrigin=headers.get("Origin");
            if(headers.getAll("Origin").size()>1||requestOrigin!=null&&!origin.equals(requestOrigin)||"cross-site".equals(headers.get("Sec-Fetch-Site")))throw new Failure(403,"仅允许本地页面访问");
            if(!Set.of("GET","POST").contains(method)) throw new Failure(405,"不支持该方法");
            if(path.contains("?")||path.contains("%")) throw new Failure(400,"不支持查询参数或编码路径");
            if(method.equals("POST")) {
                if(!origin.equals(requestOrigin)||!"1".equals(headers.get("X-Specus-UI"))) throw new Failure(403,"写入操作必须来自本地页面");
                if(!headers.get("Content-Type","").split(";",2)[0].trim().equalsIgnoreCase("application/json"))throw new Failure(415,"请求必须是 JSON");
            }
            if(Set.of("/","/app.js","/app.css").contains(path)&&method.equals("GET")) {
                String name=path.equals("/")?"index.html":path.substring(1);
                try(var stream=LocalUi.class.getResourceAsStream("/local-ui/"+name)) {
                    if(stream==null) throw new Failure(404,"页面资源缺失");result=stream.readAllBytes();
                }
                type=path.endsWith(".js")?"text/javascript; charset=utf-8":path.endsWith(".css")?"text/css; charset=utf-8":"text/html; charset=utf-8";
            } else {
                JsonNode body=UiConfig.JSON.createObjectNode();
                if(method.equals("POST")) {
                    try { body=CliOutput.JSON.readTree(bytes);if(body==null||!body.isObject())throw new IllegalArgumentException(); }
                    catch(Exception error) {throw new Failure(400,"无效 JSON 请求");}
                }
                Object data;
                if(path.equals("/api/session")&&method.equals("POST")) data=session(body);
                else {authenticate(headers.get("Authorization"));data=api(path,method,body);}
                result=CliOutput.JSON.writeValueAsBytes(data);
            }
        } catch(Exception error) {
            status=error instanceof Failure failure?failure.status:422;
            String message=error instanceof Failure?error.getMessage():"操作失败，请检查配置、权限和本地状态；原始诊断不会返回浏览器";
            try {result=CliOutput.JSON.writeValueAsBytes(Map.of("schemaVersion",1,"error",message));}
            catch(Exception impossible) {result="{}".getBytes(StandardCharsets.UTF_8);}
        }
        var response=new DefaultFullHttpResponse(HttpVersion.HTTP_1_1,HttpResponseStatus.valueOf(status),Unpooled.wrappedBuffer(result));
        response.headers().set("Content-Type",type).set("Content-Length",result.length).set("Connection","close")
                .set("Cache-Control","no-store").set("X-Content-Type-Options","nosniff").set("Referrer-Policy","no-referrer")
                .set("X-Frame-Options","DENY").set("Permissions-Policy","camera=(), microphone=(), geolocation=()")
                .set("Content-Security-Policy","default-src 'none'; script-src 'self'; style-src 'self'; connect-src 'self'; img-src 'self'; base-uri 'none'; frame-ancestors 'none'; form-action 'none'");
        ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE);
    }
    public static int run(ClientCli.Options options,String version) {
        System.setOut(new java.io.PrintStream(new java.io.FileOutputStream(java.io.FileDescriptor.out),true,StandardCharsets.UTF_8));
        // No raw core log lines (tokens/session IDs) in a browser-management terminal.
        if(org.slf4j.LoggerFactory.getLogger(org.slf4j.Logger.ROOT_LOGGER_NAME) instanceof ch.qos.logback.classic.Logger logger)
            logger.setLevel(ch.qos.logback.classic.Level.OFF);
        Path lock=null;Channel listener=null;boolean ownsLock=false;String stage="私有状态目录";
        var group=new MultiThreadIoEventLoopGroup(2,NioIoHandler.newFactory());
        var workers=new ThreadPoolExecutor(0,16,30,TimeUnit.SECONDS,new SynchronousQueue<>(),r -> {var t=new Thread(r,"local-ui-http");t.setDaemon(true);return t;});
        var ui=new LocalUi(options,version);var finished=new CountDownLatch(1);
        var connections=new java.util.concurrent.atomic.AtomicInteger();
        Thread hook=new Thread(finished::countDown,"local-ui-shutdown");
        try {
            lock=CliState.ensureRoot().resolve(CliState.prefix(options.config().toString()).substring(5)+"ui.lock");
            stage="同配置管理锁（先确认原管理进程已退出，再清理遗留锁）";
            Files.writeString(lock,Long.toString(ProcessHandle.current().pid()),StandardOpenOption.CREATE_NEW,StandardOpenOption.WRITE);
            ownsLock=true;
            stage="本机端口（检查 --port 是否被占用，或使用 --port 0）";
            listener=new ServerBootstrap().group(group).channel(NioServerSocketChannel.class).childHandler(new ChannelInitializer<SocketChannel>() {
                @Override protected void initChannel(SocketChannel channel) {
                    channel.pipeline().addLast(new ChannelInboundHandlerAdapter() {
                        boolean admitted;
                        io.netty.util.concurrent.ScheduledFuture<?> deadline;
                        @Override public void channelActive(ChannelHandlerContext ctx) {
                            if(connections.incrementAndGet()>16) {connections.decrementAndGet();ctx.close();return;}
                            admitted=true;deadline=ctx.executor().schedule(() -> {ctx.close();},25,TimeUnit.SECONDS);ctx.fireChannelActive();
                        }
                        @Override public void channelInactive(ChannelHandlerContext ctx) {
                            if(admitted) connections.decrementAndGet();if(deadline!=null)deadline.cancel(false);ctx.fireChannelInactive();
                        }
                    });
                    channel.pipeline().addLast(new ReadTimeoutHandler(10),new HttpServerCodec(4096,16384,8192),new HttpObjectAggregator(65536),new SimpleChannelInboundHandler<FullHttpRequest>() {
                        @Override protected void channelRead0(ChannelHandlerContext ctx,FullHttpRequest request) {
                            if(!request.decoderResult().isSuccess()) {ctx.close();return;}
                            byte[] body=new byte[request.content().readableBytes()];request.content().readBytes(body);
                            String path=request.uri(),method=request.method().name();HttpHeaders headers=request.headers().copy();
                            try {workers.execute(() -> ui.handle(ctx,path,method,headers,body));}
                            catch(RejectedExecutionException ignored) {ctx.close();}
                        }
                        @Override public void exceptionCaught(ChannelHandlerContext ctx,Throwable error) {ctx.close();}
                    });
                }
            }).bind(new InetSocketAddress("127.0.0.1",options.uiPort())).sync().channel();
            ui.origin="http://127.0.0.1:"+((InetSocketAddress)listener.localAddress()).getPort();
            stage="管理服务";
            System.out.println("本地管理页面："+ui.origin+"\n配置："+options.config()+"\nCtrl+C 退出管理服务及本页连接。");ui.printCode();
            if(!options.noOpen()) {
                String url=ui.origin+"/#"+ui.code;
                try {
                    String os=System.getProperty("os.name");
                    new ProcessBuilder(os.startsWith("Windows")?List.of("rundll32","url.dll,FileProtocolHandler",url):List.of(os.startsWith("Mac")?"open":"xdg-open",url))
                            .redirectOutput(ProcessBuilder.Redirect.DISCARD).redirectError(ProcessBuilder.Redirect.DISCARD).start();
                } catch(Exception error) {System.err.println("未能打开浏览器，请手动打开本地地址并输入连接码。");}
            }
            Thread.ofVirtual().start(() -> {try(var reader=new java.io.BufferedReader(new java.io.InputStreamReader(System.in))) {while(reader.readLine()!=null) ui.printCode();} catch(Exception ignored) { }});
            // JVM shutdown hooks must finish cleanup themselves, not merely wake main (which JVM may abandon).
            Path ownedLock=lock;Channel ownedListener=listener;
            hook=new Thread(() -> {ownedListener.close().syncUninterruptibly();synchronized(ui.operations) {try {ui.runtime.stop();Files.deleteIfExists(ownedLock);}catch(Exception ignored) { }}finished.countDown();},"local-ui-shutdown");
            Runtime.getRuntime().addShutdownHook(hook);finished.await();return 0;
        } catch(Exception error) {
            System.err.println("无法启动本地页面："+stage+"。请检查对应资源及当前用户权限。");return 2;
        } finally {
            if(listener!=null)listener.close().syncUninterruptibly();workers.shutdownNow();group.shutdownGracefully(0,2,TimeUnit.SECONDS).syncUninterruptibly();
            try {Runtime.getRuntime().removeShutdownHook(hook);}catch(IllegalStateException ignored) { }
            try {ui.runtime.stop();}catch(Exception ignored) { }
            // CREATE_NEW may fail: never remove another process's lock.
            if(ownsLock&&lock!=null)try {Files.deleteIfExists(lock);}catch(Exception ignored) { }
        }
    }
}
